index.js

```diff
@@ -23,24 +20,26 @@ export default (sbp('sbp/selectors/register', {
       eventQueues[name] = { events: [] }
     }
     const events = eventQueues[name].events
-    events.push({
+    const thisEvent = {
       sbpInvocation,
-      state: STATE_PENDING,
       promise: null
-    })
+    }
+    events.push(thisEvent)
     while (events.length > 0) {
       const event = events[0]
-      if (event.state === STATE_PENDING) {
-        event.state = STATE_INVOKING
-        event.promise = sbp(...event.sbpInvocation)
-        await event.promise
-        event.state = STATE_FINISHED
-      } else if (event.state === STATE_INVOKING) {
+      if (event === thisEvent) {
+        try {
+          const promise = sbp(...event.sbpInvocation)
+
+          event.promise = promise instanceof Promise ? new Promise((accept) => promise.finally(accept)) : Promise.resolve()
+
+          return await promise
+        } finally {
+          events.shift()
+        }
+      } else {
         // wait for invocation to finish
         await event.promise
```

Comment:
OK, so here's a scenario I'm concerned about with this, and it's kinda hard to describe but I'll do my best.

We start with:

```js
    const thisEvent = {
      sbpInvocation,
      promise: null
    }
    events.push(thisEvent)
```

Now we - the caller, can check this here in the loop:

```js
if (event === thisEvent) {
```

But somebody else has no ability to do this. I think it's possible that the following situation can occur:

1. We call queueEvent and wait
2. Someone else calls queueEvent and also waits
3. The scheduler gets them to reach the end of the queue first, and so they grab our event before we do and await on it. However, at this point, `await event.promise` immediately returns because it evaluates to `null`
4. They reach their event and immediately evaluate it and wait on it.
5. We reach our event and immediately evaluate it and wait on it.

Now the safety and guarantees of the event queue have been broken, because not only are events happening out of order, but they're happening at the same time too, and things explode, and do so in extreme edge cases that are difficult to even debug.

Reply:
That's a valid point. I'm unsure how JS would handle this as it's meant to be a single thread, but I see that something like that could happen. A fix is handling it in `event.promise`. Let me quickly fix this.

Reply:
The new code should always work (provided at least that the events are added to the array in order)

Reply:
Nice... 😎